solr/src/main/java/org/apache/metamodel/solr/SolrDataContext.java

```diff
@@ -0,0 +1,429 @@
+import org.apache.metamodel.query.LogicalOperator;
+import org.apache.metamodel.query.OperatorType;
+import org.apache.metamodel.query.FunctionType;
+import org.apache.metamodel.schema.Column;
+import org.apache.metamodel.schema.ColumnType;
+import org.apache.metamodel.schema.MutableColumn;
+import org.apache.metamodel.schema.MutableSchema;
+import org.apache.metamodel.schema.MutableTable;
+import org.apache.metamodel.schema.Schema;
+import org.apache.metamodel.schema.Table;
+import org.apache.metamodel.util.SimpleTableDef;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.apache.solr.client.solrj.SolrQuery.ORDER;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.impl.HttpSolrServer;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.client.solrj.response.FacetField;
+import org.apache.solr.common.SolrDocumentList;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+
+/**
+ * DataContext implementation for ElasticSearch analytics engine.
+ * 
+ * ElasticSearch has a data storage structure hierarchy that briefly goes like
+ * this:
+ * <ul>
+ * <li>Index</li>
+ * <li>Document type (short: Type) (within an index)</li>
+ * <li>Documents (of a particular type)</li>
+ * </ul>
+ * 
+ * When instantiating this DataContext, an index name is provided. Within this
+ * index, each document type is represented as a table.
+ * 
+ * This implementation supports either automatic discovery of a schema or manual
+ * specification of a schema, through the {@link SimpleTableDef} class.
+ */
+
+public class SolrDataContext extends QueryPostprocessDataContext implements
+		DataContext {
+
+	private static final Logger logger = LoggerFactory
+			.getLogger(SolrDataContext.class);
+
+	public static final String FIELD_ID = "_id";
+
+	private final SimpleTableDef tableDef;
+	private final String indexName;
+	private final String url;
+
+	private SolrQuery query = new SolrQuery();
+
+	private static enum queryTypes {
+		ROW, FACET
+	};
+
+	private static final int MAX_RETRIES = 0;
+	private static final int SOCK_TIMEOUT = 1000;
+	private static final int CONN_TIMEOUT = 5000;
+	private static final int DEFAULT_LIMIT = 1000;
+
+	public SolrDataContext(String url, String indexName) {
+		this.url = url;
+		this.indexName = indexName;
+		this.tableDef = detectSchema(url, indexName);
+	}
+
+	private SimpleTableDef detectSchema(String url, String indexName) {
+		SimpleTableDef tableDef = null;
+
+		url += "/schema";
+
+		try {
+			HttpClient httpclient = new DefaultHttpClient();
+			HttpGet request = new HttpGet(url);
+			HttpResponse response = httpclient.execute(request);
+
+			InputStreamReader inputStream = new InputStreamReader(response
+					.getEntity().getContent());
+
+			ObjectReader reader = new ObjectMapper().reader(HashMap.class);
```

Comment:
Is there not a Solr client API for doing these things? If not, OK, but to me it seems like a lot of work to do the JSON parsing manually - I could imagine many things going wrong with this if the response is not quite like we expect it.